modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/service/impl/DDMStructureLocalServiceImpl.java

```diff
@@ -2289,6 +2424,35 @@ private void _validate(
 		}
 	}
 
+	private void _validate(
+			String externalReferenceCode, long groupId, long parentStructureId,
+			long classNameId, String structureKey, Map<Locale, String> nameMap,
+			DDMForm ddmForm)
+		throws PortalException {
+
+		_validateExternalReferenceCode(
+			externalReferenceCode, groupId, classNameId);
+
+		_validate(
```

Comment:
Since only one `addStructure` method needs to call `_validateExternalReferenceCode`, maybe we can move this validation there. Am I right? What do you think?

Reply:
Makes sense! I create this method to avoid future complications, but turns out the other _add  (for this case, when duplicating/copying a DDMStructure) don't need to worry, as I'm generating a new,random ERC [here](https://github.com/liferay-bpm/liferay-portal/pull/3794/files/c3526c647ee131d92d6f1967ddf21962e447dc5f..2bfcd044bc955980ad446f3960166feced8fae28#diff-16ae57c3551c581526e9e38466853c78ab9b5afde254ff07ec39b8dbf529a2d0R425).